src/coprocessor/cache.rs

```diff
@@ -0,0 +1,491 @@
+        self.result = res;
+        self.region_id = region_id;
+        self.size - old_size
+    }
+}
+
+// RegionDistSQLCacheEntry track Region's cache items key.
+// So we can quickly get a cache entry list by given a region id.
+// That can make evict_region operation more faster than scan whole
+// cache items and get the remove item list.
+pub struct RegionDistSQLCacheEntry {
+    version: u64,
+    enable: bool,
+    cached_items: HashMap<DistSQLCacheKey, u8>,
+}
+
+impl Default for RegionDistSQLCacheEntry {
+    fn default() -> RegionDistSQLCacheEntry {
+        RegionDistSQLCacheEntry {
+            version: 0,
+            enable: true,
+            cached_items: HashMap::default(),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct DistSQLCache {
+    regions: HashMap<u64, RegionDistSQLCacheEntry>,
+    max_size: usize,
+    map: LinkedHashMap<DistSQLCacheKey, DistSQLCacheEntry>,
+    size: usize,
+}
+
+impl DistSQLCache {
+    // capacity is memory size unit is byte
+    pub fn new(capacity: usize) -> DistSQLCache {
+        DistSQLCache {
+            max_size: capacity,
+            ..Default::default()
+        }
+    }
+
+    pub fn put(&mut self, region_id: u64, k: DistSQLCacheKey, version: u64, res: Vec<u8>) {
+        if !self.is_region_cache_enabled(region_id) {
+            return;
+        }
+
+        let key_size = k.len();
+        if self.map.contains_key(&k) {
+            let mut entry = self.map.get_mut(&k).unwrap();
+            let size_diff = entry.update(region_id, version, key_size, res);
+            self.size += size_diff;
+        } else {
+            let entry = DistSQLCacheEntry::new(region_id, version, key_size, res);
+            self.size += entry.size;
+            self.map.insert(k.clone(), entry);
+            self.update_regions(region_id, k);
+        }
+
+        // Remove entry untile cache size is less or equals than capacity
+        while self.size() > self.capacity() {
+            self.remove_lru();
+        }
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["size"])
+            .set(self.size() as f64);
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["count"])
+            .set(self.len() as f64);
+    }
+
+    fn check_evict_key(&mut self, region_id: u64, k: &str) {
+        let mut need_remove = false;
+        if let Some(entry) = self.map.get(k) {
+            if let Some(rentry) = self.regions.get(&region_id) {
+                // Region's version is not same as cached evict it
+                if entry.version != rentry.version {
+                    need_remove = true;
+                }
+            }
+        }
+        if need_remove {
+            self.remove(k);
+        }
+    }
+
+    pub fn get_region_version_and_cache_entry(
+        &mut self,
+        region_id: u64,
+        k: &str,
+    ) -> (u64, Option<&Vec<u8>>) {
+        (self.get_region_version(region_id), self.get(region_id, k))
```

Comment:
Call `get_region_version` is not good. We can get both `version` and `enable` from `self.regions[region_id]` at one time.